Add Scine QCMaquis recipe to builtin packages

adam-grofe opened the pull request:
[Scine QCMaquis](https://scine.ethz.ch/download/qcmaquis) is a DMRG program developed by the Markus Reiher group at ETH. In this PR, I add a spack recipe for compiling the code.

Commit:
Add Scine QCMaquis recipe to builtin packages

Commit:
Format scine-qcmaquis recipe using black

Commit:
Remove import os from scine-qcmaquis recipe

Commit:
Reduce length of symmetries line in scine-qcmaquis

Commit:
Add myself as a maintainer of the scine-qcmaquis recipe

tldahlgren commented on var/spack/repos/builtin/packages/scine-qcmaquis/package.py:
```diff
@@ -0,0 +1,63 @@
+
+    def cmake_args(self):
+        args = []
+        build_symm = str(self.spec.variants["symmetries"].value)
+        if build_symm:
+            args.extend(["-DBUILD_SYMMETRIES={}".format(build_symm)])
+        build_tests = str(self.spec.variants["build_tests"].value)
+        if build_tests:
+            args.extend(["-DQCMAQUIS_TESTS=On"])
```
Have you checked out the helper functions at https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake?

adam-grofe replied:
Thank you for the suggestion; I made the suggested changes.

tldahlgren commented on var/spack/repos/builtin/packages/scine-qcmaquis/package.py:
```diff
@@ -0,0 +1,63 @@
+        default="openblas",
+        description="Which blas library to use.",
+    )
+    variant(
+        "symmetries",
+        default="TwoU1;SU2U1;SU2U1PG;TwoU1PG",
+        description='Wave functions symmetries to compile (e.g. "TwoU1;SU2U1;SU2U1PG;TwoU1PG")',
+    )
```
Would it make sense for this package to make this a multi-valued variant that can take multiple values?  (See the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#multi-valued-variants.)

adam-grofe replied:
I changed this variant to use the multi.

tldahlgren commented on var/spack/repos/builtin/packages/scine-qcmaquis/package.py:
```diff
@@ -0,0 +1,63 @@
+        build_symm = str(self.spec.variants["symmetries"].value)
+        if build_symm:
+            args.extend(["-DBUILD_SYMMETRIES={}".format(build_symm)])
+        build_tests = str(self.spec.variants["build_tests"].value)
+        if build_tests:
+            args.extend(["-DQCMAQUIS_TESTS=On"])
```
This isn't doing what you want.  

Try `spack install scene-qcmaquis~build_tests` (so disable the variant) and you'll see that the option is translated to `'-DQCMAQUIS_TESTS=On'`.

adam-grofe replied:
After changing the cmake args function to use `define_from_variant`, this bug was fixed. I verified it myself in a clean environment.

tldahlgren reviewed (changes requested):
Confirmed the homepage and three version branches; however, the variants appear to need work.

Commit:
Update Scine-QCMaquis recipe following the review

PR URL: https://github.com/spack/spack/pull/39709

Changes:
  - Updated Symmetries variant to us multi feature
  - Added dependency to boost+chrono since it was undocumented
  - Use define_from_variant to setup CMake args
  - Make version 3.1.2 be preferred since 3.1.3 build is broken

adam-grofe committed:
Merge branch 'develop' into feature/scine_qcmaquis_recipe

adam-grofe committed:
Merge branch 'develop' into feature/scine_qcmaquis_recipe

tldahlgren commented on var/spack/repos/builtin/packages/scine-qcmaquis/package.py:
```diff
@@ -0,0 +1,62 @@
+        "symmetries",
+        default="SU2U1PG,TwoU1PG",
+        description='Wave functions symmetries to compile (e.g. "SU2U1PG,TwoU1PG")',
+        values=("U1", "TwoU1", "TwoU1PG", "NU1", "Z2", "SU2U1", "SU2U1PG", "U1DG", "NONE"),
```
Should the value(s) be validated?  For example, what if someone included `NONE`?

adam-grofe replied:
According the QcMaquis CMakeLists.txt, "TwoU1PG and SU2U1PG are built in any case."

tldahlgren commented on var/spack/repos/builtin/packages/scine-qcmaquis/package.py:
```diff
@@ -0,0 +1,62 @@
+    depends_on("intel-oneapi-mkl", when="blas=mkl")
+    depends_on("gsl")
+    depends_on("boost+program_options+filesystem+system+thread+serialization+chrono")
+    depends_on("googletest+gmock", when="build_tests=True")
```
While this works, conditional dependencies involving boolean variants usually use the `+` or `~` syntax (i.e., `when="+build_tests"`) as shown in the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-dependencies.

adam-grofe replied:
Thank you for the suggestion. I have made this change to the package.

tldahlgren reviewed (commented):
Looking good.  Just have a couple of questions/pointers.

tldahlgren commented:
Please ping me when you're ready for this to be merged.

Commit:
Change build_tests boolean condition

adam-grofe committed:
Merge branch 'develop' into feature/scine_qcmaquis_recipe

adam-grofe committed:
Merge branch 'develop' into feature/scine_qcmaquis_recipe

adam-grofe committed:
Merge branch 'develop' into feature/scine_qcmaquis_recipe
